Services/Component/classes/class.ilPluginAdmin.php

```diff
@@ -57,102 +59,200 @@ function __construct()
+				$inactive_reason = "Update needed.";
 			}
-			else if ($rec["last_update_version"] != $version)
+			$plugin_data["inactive_reason"] = $inactive_reason;
+			$plugin_data["needs_update"] = true;
+			$plugin_data["activation_possible"] = false;
+		}
+		else if (ilComponent::isVersionGreaterString($last_update_version, $plugin_data["version"]))
+		{
+			$plugin_data["is_active"] = false;
+			if (is_object($this->lng))
 			{
-				$active = false;
-				if (is_object($lng))
-				{
-					$inactive_reason = $lng->txt("cmps_needs_update");
-				}
-				else
-				{
-					$inactive_reason = "Update needed.";
-				}
-				$needs_update = true;
-				$activation_possible = false;
+				$inactive_reason = $this->lng->txt("cmps_needs_upgrade");
+			}
+			else
+			{
+				$inactive_reason = "Upgrade needed.";
+			}
+			$plugin_data["inactive_reason"] = $inactive_reason;
+			$plugin_data["activation_possible"] = false;
+		}
+		else if ($last_update_version != $plugin_data["version"])
+		{
+			$plugin_data["is_active"] = false;
+			if (is_object($this->lng))
+			{
+				$inactive_reason = $this->lng->txt("cmps_needs_update");
 			}
+			else
+			{
+				$inactive_reason = "Update needed.";
+			}
+			$plugin_data["inactive_reason"] = $inactive_reason;
+			$plugin_data["needs_update"] = true;
+			$plugin_data["activation_possible"] = false;
+		}
+	}
 
-			$this->data[$a_ctype][$a_cname][$a_slot_id][$a_pname]["is_active"] = $active;
-			$this->data[$a_ctype][$a_cname][$a_slot_id][$a_pname]["inactive_reason"] = $inactive_reason;
-			$this->data[$a_ctype][$a_cname][$a_slot_id][$a_pname]["needs_update"] = $needs_update;
-			$this->data[$a_ctype][$a_cname][$a_slot_id][$a_pname]["activation_possible"] = $activation_possible;
+	/**
+	 * Set plugin data for intall
+	 *
+	 * @param string[] 	&$plugin_data
+	 *
+	 * @return null
+	 */
+	protected function setMustInstall(array &$plugin_data) {
+		$plugin_data["must_install"] = true;
+		$plugin_data["is_active"] = false;
+		$plugin_data["needs_update"] = false;
+		$plugin_data["activation_possible"] = false;
 
-			$this->got_data[$a_ctype][$a_cname][$a_slot_id][$a_pname] = true;
+		if (is_object($this->lng))
+		{
+			$inactive_reason = $this->lng->txt("cmps_must_installed");
 		}
+		else
+		{
+			$inactive_reason = "Plugin must be installed.";
+		}
+		$plugin_data["inactive_reason"] = $inactive_reason;
+	}
+
+	/**
+	 * Set plugin data according current db state
+	 *
+	 * @param string[] &$plugin_data
+	 * @param bool 	$active
+	 *
+	 * @return null
+	 */
+	protected function setCurrentState(array &$plugin_data, $active) {
+		$plugin_data["is_active"] = $active;
+		$plugin_data["activation_possible"] = !$active;
+		$plugin_data["must_install"] = false;
+		$plugin_data["needs_update"] = false;
+		$plugin_data["inactive_reason"] = "";
+	}
+
+	/**
+	 * Get informations from plugin php file
+	 *
+	 * @param string 	$plugin_php_file
+	 *
+	 * @return string[]
+	 */
+	protected function parsePluginPhp($plugin_php_file) {
```

Comment:
I find this plugin.php file horrible, but I'm not sure that we really want to replace the proven PHP-internal parsing (i.e. `include`) with this just now. Is there another reason you introduced this in here?